packages/x-adapter/src/endpoint-adapter.factory.ts

```diff
@@ -0,0 +1,74 @@
+import { fetchHttpClient } from './http-clients/fetch.http-client';
+import { identityMapper } from './mappers/identity.mapper';
+import {
+  EndpointAdapter,
+  EndpointAdapterFactory,
+  EndpointAdapterOptions
+} from './types/adapter.types';
+import { Mapper } from './types/mapper.types';
+import { interpolate } from './utils/interpolate';
+
+/**
+ * Factory to create {@link EndpointAdapter | endpoint adapters} with the given
+ * {@link EndpointAdapterOptions | options}.
+ *
+ * @param options - The {@link EndpointAdapterOptions | options} to create a new
+ * {@link EndpointAdapter} with.
+ *
+ * @returns A brand new {@link EndpointAdapter} object.
+ * @public
+ */
+export const endpointAdapterFactory: EndpointAdapterFactory = <Request, Response>(
+  options: EndpointAdapterOptions<Request, Response>
+) => {
+  const endpointAdapter: EndpointAdapter<Request, Response> = (
+    request,
+    { endpoint: requestEndpoint, ...requestOptions } = {}
+  ) => {
+    const {
+      endpoint: rawEndpoint,
+      httpClient = fetchHttpClient,
+      requestMapper = identityMapper,
+      responseMapper = identityMapper
+    } = options;
+
+    /**
+     * Returns an endpoint.
+     *
+     * @param endpoint - The endpoint to process.
+     * @param request - The request object.
+     *
+     * @returns The endpoint.
+     * @internal
+     */
+    function getEndpoint<Request>(
+      endpoint: string | Mapper<Request, string> | undefined,
+      request: Request
+    ): string {
+      if (!endpoint) {
+        throw Error('Tried to make a request without an endpoint');
+      }
+      return typeof endpoint === 'function'
+        ? endpoint(request, {})
+        : interpolate(endpoint, request as unknown as Record<string, string | number | boolean>);
```

Comment:
Modify also the `interpolate` function so it accepts `null` and `undefined` values. With this cast we are allowing `null` or `undefined` values to be passed, which will cause an error:

```ts
export function interpolate(
  string: string,
  parameters: Record<string, unknown>
): string {
```

Probably just ignoring nullish values is enough:

```ts
        parameters[property] != null ? `${head}${String(parameters[property])}${tail}` : ''
```

Reply:
Thanks for the refactor. Done